### ISSUE TITLE
Issue with "Array.prototype.sort()": (short summary here please)

### ISSUE THREAD

**Issue by JoshuvaGeorge03** [mdn/content#2769]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description

#### What information was incorrect, unhelpful, or incomplete?

The below statement is incorrect.

If compareFunction(a, b) returns less than 0, leave a and b unchanged.

#### Specific section or headline?
In the **description** section of the sort, you can find this.

#### What did you expect to see?
The correct definition is, sort a before b

#### Did you test this? If so, how?

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/array/sort`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/array/sort/index.html
* Last commit: https://github.com/mdn/content/commit/4f341c4f28f35ed21e56c96e44d32879a0baeae9
* Document last modified: 2021-02-25T12:49:25.000Z

</details>

**Comment by CreaTorAlexander** [comment]
But to sort a before b when a is before b means that it will leave it unchanged

**Comment by JoshuvaGeorge03** [comment]
Thanks for the explanation. 

Yes, I do understand your explanation.

But As a novice developer, when someone sees this explanation and Then can easily be get confused with the returning 0.

`
If compareFunction(a, b) returns less than 0, leave a and b unchanged.
If compareFunction(a, b) returns 0, leave a and b unchanged with respect to each other
`

Please correct me if I am wrong.

**Comment by CreaTorAlexander** [comment]
Yes I understand this point and I agree. On the other hand when we will update it there will probably be people getting confused why we don't just write that it will leave unchanged.

It is worth to discuss what the better choice here is.

**Comment by JoshuvaGeorge03** [comment]
Sorry for the late reply.

I would suggest, adding some kind of explanation before the statement (`If compareFunction(a, b) returns 0, leave a and b unchanged with respect to each other`)  to say how it differs from this statement (`If compareFunction(a, b) returns 0, leave a and b unchanged with respect to each other');

I want to point out one more thing, Before MDN docs said `If compareFunction(a, b) returns less than 0, sort a before b`, something like that.

**Comment by CreaTorAlexander** [comment]
All good @JoshuvaGeorge03, that is a good idea, I will take a look at it later on :+1:

**Comment by JoshuvaGeorge03** [comment]
@CreaTorAlexander Thanks.

**Comment by himanshugarg** [comment]
@CreaTorAlexander , is it not incorrect to say - **sort will leave `a` and `b` unchanged if `compareFunction(a, b)` returns less than 0**? What if `b` actually appears before `a` in the array? Then sort will switch their relative order? Or are we _implying_ that `compareFunction` is only called such that `a` appears before `b`?

**Comment by hamishwillee** [comment]
@himanshugarg Re ^^^ No it is not incorrect: it is the _definition_ of` compareFunction(a, b)` that if the value returns less than zero the sort algorithm will not change the order of any two compared items.

**Comment by hamishwillee** [comment]
I believe I have fixed this in https://github.com/mdn/content/pull/5224 (@CreaTorAlexander Sorry to steal this one off you - I figure if we have this many questions, it needs to be addressed :-) ).

@JoshuvaGeorge03 @CreaTorAlexander Please check it out and please confirm it removes any ambiguity or confusion for you.

**Comment by JoshuvaGeorge03** [comment]
@hamishwillee I understand the new definition, but have a few questions. please answer

From your definition, returning 0 and less than 0 do the same thing. Please correct me if I am wrong.

if so, can we  interchange both of them? is it will work on those scenarios?

What I mean by this is, can we able to return 0 for less than condition also and can we return -1 for equality?

**Comment by hamishwillee** [comment]
> if so, can we interchange both of them? is it will work on those scenarios?
> What I mean by this is, can we able to return 0 for less than condition also and can we return -1 for equality?

My guess is that if you return -1 for equality you probably remove the guarantee that the two equal compared items will not be re-ordered (i.e. that it is a "stable sort"). That's a guess - but generally speaking I wouldn't try and second guess the spec authors :-)

**Comment by eriktate** [comment]
Sorry to piggy back on a closed issue, but it looks like there's been a couple of duplicate issues already created.

Based on a few discussions I've been able to find, with folks who know way more about how sort is implemented than I do, I agree that the definition of `compareFunction` is technically correct. Meaning that return values <= 0 don't result in any order changes for elements `a` and `b`. However, I think that the definition is lacking some information (or warnings?) about inconsistent vs. consistent compare functions. In particular, the line `If compareFunction(a, b) returns a value ≤ 0, leave a and b in the same order.` implies that there's fundamentally no difference between returning `0` or `-1`. By that logic, these compare functions should yield the same results (assuming numerical values):
```javascript
const inconsistentCompare = (a, b) => a > b ? 1 : 0; // inconsistent because we never return negative values
const consistentCompare = (a, b) => a - b;
```
In practice, though, these can (and usually) yield entirely different results. Again, I don't think the docs are incorrect, I think it's just not necessarily obvious to a reader that it's important that you specifically return 0 _only_ for items that are equal, and that you return some negative number _only_ for items that are considered less than. Or more succinctly, I think it's important that reader understands what a consistent compare function is and why they should care about consistency.

I only _barely_ understand what's going on, and I didn't know about consistent vs. inconsistent compare functions until a few days ago, so I'm not sure I could provide the best verbiage for making this more clear 😅  Mostly wondering if others agree that the text as written is still confusing?

**Comment by mistval** [comment]
I commented in more detail in https://github.com/mdn/content/pull/6255 but this code:

```js
console.log(
  'Sorted array: ' + [2,1].sort((a, b) => {
    console.log(`a: ${a}`);
    console.log(`b: ${b}`);
    const compareResult = a - b;
    console.log(`Compare Result: ${compareResult}`);
    return compareResult;
  })
)
```
executed in Chrome outputs:
```js
a: 1
b: 2
Compare Result: -1
Sorted array: 1,2
```
That is, the compare function returned -1, yet the order of the elements was changed (`[2,1]` was sorted to `[1,2]`), which is inconsistent with what this documentation states.

The output in Firefox is different (`a` is 2 and `b` is 1) which leads me to believe that "If compareFunction(a, b) returns a value ≤ 0, leave a and b in the same order." is a SpiderMonkey implementation detail and not a specification requirement. Indeed I couldn't find anything in the specification that requires sort() to behave in that way specific way.
